### DISCUSSION TITLE
Debugging in Arduino IDE 2.0 with PicoProbe (Raspberry Pi Debug Probe/OpenOCD/CMSIS-DAP)

### DISCUSSION THREAD

**Discussion by MarcusGarfunkel** [earlephilhower/arduino-pico#1299] [Q&A] [answered]
Hello,
I have a Raspberry Pi Debug Probe (This is a hardware device that runs Picoprobe/OpenOCD/CMSIS-DAP). This is connected to a Raspberry Pi Pico H. I can program the Pico H using Arduino IDE 2.0. This is accomplished by selecting Tools->Program Method: Picoprobe (CMSIS-DAP) and then pressing the upload button.

My question is, how can I debug using this setup? The only option available in Arduino IDE 2.0 is the "Arduino" configuration. I don't seem to be able to create any other configurations. The output of the gdb-server is the following:

```
Waiting for gdb server to start...[2023-03-11T21:04:50.587Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session connected. You can switch to "DEBUG CONSOLE" to see GDB interactions.
"C:\\Users\\Marcus Garfunkel\\AppData\\Local\\Arduino15\\packages\\rp2040\\tools\\pqt-openocd\\1.5.0-b-c7bab52/bin/openocd" -c "gdb_port 50002" -c "tcl_port 50003" -c "telnet_port 50004" -s "c:\\Users\\Marcus Garfunkel\\OneDrive - Governor Media LLC\\Documents\\Projects for customers\\Governor Media\\MezMirage\\Arduino Sketches\\2023-03-11 Debug Probe\\Blink-2023-03-11" -f "C:/Users/Marcus Garfunkel/AppData/Local/Programs/Arduino IDE/resources/app/plugins/cortex-debug/extension/support/openocd-helpers.tcl" -f "C:\\Users\\Marcus Garfunkel\\AppData\\Local\\Arduino15\\packages\\rp2040\\hardware\\rp2040\\2.7.3/lib/picoprobe.tcl"
Open On-Chip Debugger 0.11.0-g228ede43d-dirty (2022-12-22-20:10)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
CDRTOSConfigure
Info : only one transport option; autoselect 'swd'
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
Info : Listening on port 50003 for tcl connections
Info : Listening on port 50004 for telnet connections
Error: Failed to open or find the device
Error: Can't find a picoprobe device! Please check device connections and permissions.

[2023-03-11T21:04:50.681Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...
```

Any help would be appreciated. Feel free to reach out with any questions or concerns.

Sincerely,
Marcus

**Comment by earlephilhower** [comment, marked as answer]
Use Picoprobe-CMSIS-DAP.tcl instead.  The RPI folks made a 2nd version of the picoprobe FW based on CMSIS-DAP, which is incompatible with the original Picoprobe firmware.

**Reply by BetterComputing** [reply]
There seems to be something else wrong here. I am using Linux Mint and I've installed your Board Manager through the Arduino IDE. I also have a Pico with the CMSIS-DAP firmware running on it and connected to my target board. I selected **Board: Generic RP2040** and **Upload Method: Picoprobe (CMSIS-DAP)**. When I try to debug, the command line issued by the Arduino IDE uses **picoprobe.tcl**, not **picoprobe_cmsis_dap.tcl**. I manually edited the **boards.txt** on my system and found that no matter which Upload Method I select, the values from **Default (UF2)** are used. If I replace **picoprobe.tcl** with **picoprobe_cmsis_dap.tcl** in the default section of boards.txt, it works. So it seems that the section "generic.menu.uploadmethod.default" is always used regardless of the selected Upload Method.

As an aside, there are some artifacts from your local system that have been included in the Board Manager. For example, when I paused my program and the code stopped in tu_fifo_read(), an error appeared:
`702	/home/earle/Arduino/hardware/pico/rp2040/pico-sdk/lib/tinyusb/src/common/tusb_fifo.c: No such file or directory.`
For my system it should have looked for
`/home/user/.arduino15/packages/rp2040/hardware/rp2040/3.1.0/pico-sdk/lib/tinyusb/src/common/tusb_fifo.c`.

Thanks. by the way, for the Board Manager for RP2040 - awesome.

**Reply by earlephilhower** [reply]
> ... As an aside, there are some artifacts from your local system ...

That's just the way GCC debug works.  It's recording the location of the file as it was compiled.  There is a GCC substitute path command you can use to have it search/replace the path from my system to your own.  See (for example) https://www-zeuthen.desy.de/unix/unixguide/infohtml/gdb/Source-Path.html

**Reply by xyzzy-099** [reply]
This is marked as a "solution", but I have this problem, and it does not appear to be possible to change the device used.  I have 'upload method' set to "Picoprobe (CMSIS-DAP), but the Arduino 2 IDE still insists on using 'picoprobe.tcl' instead of 'picoprobe_cmsis_dap.tcl'.  In fact, no matter what I set the upload method to, it always uses 'picoprobe.tcl'.  It's not possible to edit the 'launch.json' in the GUI (like all non-sketch files, it is read-only in the GUI).  Is there something I am missing here?

**Reply by BetterComputing** [reply]
@xyzzy-099: This is still a problem and the only "solution" I have found is to copy contents of **picoprobe_cmsis_dap.tcl** to **picoprobe.tcl** (Location on Linux: **~/.arduino15/packages/rp2040/hardware/rp2040/3.1.1/lib**, Location on Windows: **C:\Users\{username}\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.1.1\lib**). Note that whenever the RP2040 library is updated the manual copy that you performed will be overwritten and you'll have to do it again.

I suspect the problem is that the Arduino IDE is incorrectly matching the list item '**Picoprobe**' to the selected list item '**Picoprobe (CMSIS-DAP)**'  because it gives up the search at the space.

**Reply by xyzzy-099** [reply]
Ok, that makes sense...  Maybe the name could just be changed to CMSIS_DAP (w/o) the picoprobe part?  I guess the right fix is to get the Arduino IDE devs to fix the matching, though.  At least I now have a workaround, thank you very much :)

**Comment by victoryred** [comment]
Starting the GDB server yields- Request 2 cancelled on connection close

I'm using the Arduino 2.2.1 IDE with dual RPi Pico's under cmsis-dap.

I'm including a copy of the output and Debug Console...for reference.

OUTPUT Window:
Open On-Chip Debugger 0.12.0-g4d87f6dca (2023-09-11-23:12)

Licensed under GNU GPL v2

For bug reports, read

http://openocd.org/doc/doxygen/bugs.html

Info : Hardware thread awareness created

Info : Hardware thread awareness created

adapter speed: 5000 kHz

Info : Using CMSIS-DAPv2 interface with VID:PID=0x2e8a:0x000c, serial=E661640843626C27

Info : CMSIS-DAP: SWD supported

Info : CMSIS-DAP: Atomic commands supported

Info : CMSIS-DAP: Test domain timer supported

Info : CMSIS-DAP: FW Version = 2.0.0

Info : CMSIS-DAP: Interface Initialised (SWD)

Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0

Info : CMSIS-DAP: Interface ready

Info : clock speed 5000 kHz

Info : SWD DPIDR 0x0bc12477, DLPIDR 0x00000001

Info : SWD DPIDR 0x0bc12477, DLPIDR 0x10000001

Info : [rp2040.core0] Cortex-M0+ r0p1 processor detected

Info : [rp2040.core0] target has 4 breakpoints, 2 watchpoints

Info : [rp2040.core1] Cortex-M0+ r0p1 processor detected

Info : [rp2040.core1] target has 4 breakpoints, 2 watchpoints

Info : starting gdb server for rp2040.core0 on 3333

Info : Listening on port 3333 for gdb connections

[rp2040.core0] halted due to debug-request, current mode: Thread

xPSR: 0xf1000000 pc: 0x000000ea msp: 0x20041f00

[rp2040.core1] halted due to debug-request, current mode: Thread

xPSR: 0xf1000000 pc: 0x000000ea msp: 0x20041f00

** Programming Started **

Info : Found flash device 'win w25q16jv' (ID 0x001540ef)

Info : RP2040 B0 Flash Probe: 2097152 bytes @0x10000000, in 32 sectors

Info : Padding image section 1 at 0x10013b6c with 148 bytes (bank write end alignment)

Warn : Adding extra erase range, 0x10013c00 .. 0x1001ffff

** Programming Finished **

** Verify Started **

** Verified OK **

[rp2040.core0] halted due to debug-request, current mode: Thread

xPSR: 0xf1000000 pc: 0x000000ea msp: 0x20041f00

[rp2040.core1] halted due to debug-request, current mode: Thread

xPSR: 0xf1000000 pc: 0x000000ea msp: 0x20041f00

Please note:
1.) I do not have any breakpoints/Watchpoints set.
2.) No idea why the GDB server was started prior to flash programming.
3.) Programming completed successfully.
4.) Looks like GDB tried to restart but halted again with the following Notification- Request 2 cancelled on connection close.

**Comment by BennetLevine1** [comment]
It is working for me. I suppose they have had time to fix the issues. I don't believe I installed anything special to get this to work, however, I did install a number of USB related libraries for my application. I am running Arduino 2.3.4 and using the Raspberry Pi debug probe. I used serial for the debug probe and for the upload method I used PicoProbe. I am also using port COM3 which it offered to me. I am not using the UART port on the debugger. Make sure you set optimize to debug. I can set breakpoints and look at my variable values when execution is paused at a breakpoint. I can also single step. This is very nice for such a low-cost device and debugger.  If you can see the picture below you can see my settings. Press the check mark symbol near the top left to compile the sketch, the right pointing arrow symbol to upload your compiled file to the pico and press the "play" symbol to start the debugger.
![arduino_debugging](https://github.com/user-attachments/assets/bf8a115b-0045-434a-93c9-a3cc3149ce14)
